pkg/storage/swift/swift.go

```diff
@@ -1,84 +1,223 @@
 	Listers *regopclient.Listers
 }
 
+// replaceEmpty is a helper function to replace empty fields with another field
+func replaceEmpty(a string, b string) string {
+	if a == "" {
+		return b
+	}
+	return a
+}
+
+// getSwiftClient returns a client that allows to interact with the OpenStack Swift service
+func (d *driver) getSwiftClient(cr *imageregistryv1.Config) (*gophercloud.ServiceClient, error) {
+	var username, password string
+	cfg, err := clusterconfig.GetSwiftConfig(d.Listers)
+	if err != nil {
+		return nil, err
+	}
+
+	// In a case of IPI all credentials are stored in clouds.yaml file in the secret, prepared by
+	// cloud-credentials-operator, and we read them from cfg.Storage.Swift.Cloud.
+	// For UPI it's enough to provide username and password in the related secret.
+	if cfg.Storage.Swift.Cloud != nil {
+		cloud := cfg.Storage.Swift.Cloud
+
+		d.Config.AuthURL = replaceEmpty(d.Config.AuthURL, cloud.AuthInfo.AuthURL)
+		d.Config.Tenant = replaceEmpty(d.Config.Tenant, cloud.AuthInfo.ProjectName)
+		d.Config.TenantID = replaceEmpty(d.Config.TenantID, cloud.AuthInfo.ProjectID)
+		d.Config.Domain = replaceEmpty(d.Config.Domain, cloud.AuthInfo.DomainName)
+		d.Config.DomainID = replaceEmpty(d.Config.DomainID, cloud.AuthInfo.DomainID)
+		d.Config.Domain = replaceEmpty(d.Config.Domain, cloud.AuthInfo.UserDomainName)
+		d.Config.DomainID = replaceEmpty(d.Config.DomainID, cloud.AuthInfo.UserDomainID)
+		d.Config.RegionName = replaceEmpty(d.Config.RegionName, cloud.RegionName)
+
+		username = cloud.AuthInfo.Username
+		password = cloud.AuthInfo.Password
+	} else {
+		username = cfg.Storage.Swift.Username
+		password = cfg.Storage.Swift.Password
+	}
+
+	opts := &gophercloud.AuthOptions{
+		IdentityEndpoint: d.Config.AuthURL,
+		Username:         username,
+		Password:         password,
+		DomainID:         d.Config.DomainID,
+		DomainName:       d.Config.Domain,
+		TenantID:         d.Config.TenantID,
+		TenantName:       d.Config.Tenant,
+	}
+
+	provider, err := openstack.AuthenticatedClient(*opts)
+	if err != nil {
+		return nil, err
+	}
+
+	endpointOpts := gophercloud.EndpointOpts{
+		Region: d.Config.RegionName,
+	}
+
+	client, err := openstack.NewContainerV1(provider, endpointOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+// NewDriver creates new Swift driver for the Image Registry
 func NewDriver(c *imageregistryv1.ImageRegistryConfigStorageSwift, listers *regopclient.Listers) *driver {
 	return &driver{
 		Config:  c,
 		Listers: listers,
 	}
 }
 
+// Secrets returns a map of the storage access secrets
 func (d *driver) Secrets() (map[string]string, error) {
+	cfg, err := clusterconfig.GetSwiftConfig(d.Listers)
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.Storage.Swift.Cloud == nil {
+		return map[string]string{
+			"REGISTRY_STORAGE_SWIFT_USERNAME": cfg.Storage.Swift.Username,
+			"REGISTRY_STORAGE_SWIFT_PASSWORD": cfg.Storage.Swift.Password,
+		}, nil
+	}
```

Comment:
I believe (and @dmage correct me if I'm wrong) we always want to publish the swift username/password in this map.

Reply:
Done